// baseui/pin-code/styled-components.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledRoot = exports.StyledInputOverrideRoot = exports.StyledInputOverrideInput = void 0;

var _styles = require("../styles");

var _input = require("../input");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var StyledRoot = (0, _styles.styled)('div', {
  display: 'flex',
  alignItems: 'center'
});
exports.StyledRoot = StyledRoot;
StyledRoot.displayName = "StyledRoot";
StyledRoot.displayName = 'StyledRoot';
var StyledInputOverrideRoot = (0, _styles.withStyle)(_input.StyledRoot, function (_ref) {
  var _SIZE$mini$SIZE$compa;

  var _ref$$size = _ref.$size,
      $size = _ref$$size === void 0 ? _input.SIZE.default : _ref$$size;
  var width = (_SIZE$mini$SIZE$compa = {}, _defineProperty(_SIZE$mini$SIZE$compa, _input.SIZE.mini, '32px'), _defineProperty(_SIZE$mini$SIZE$compa, _input.SIZE.compact, '36px'), _defineProperty(_SIZE$mini$SIZE$compa, _input.SIZE.default, '48px'), _defineProperty(_SIZE$mini$SIZE$compa, _input.SIZE.large, '56px'), _SIZE$mini$SIZE$compa)[$size];
  return {
    width: width,
    marginRight: '8px'
  };
});
exports.StyledInputOverrideRoot = StyledInputOverrideRoot;
StyledInputOverrideRoot.displayName = "StyledInputOverrideRoot";
StyledInputOverrideRoot.displayName = 'StyledInputOverrideRoot';
var StyledInputOverrideInput = (0, _styles.withStyle)(_input.StyledInput, {
  textAlign: 'center',
  paddingLeft: '0',
  paddingRight: '0'
});
exports.StyledInputOverrideInput = StyledInputOverrideInput;
StyledInputOverrideInput.displayName = "StyledInputOverrideInput";
StyledInputOverrideInput.displayName = 'StyledInputOverrideInput';
