// baseui/tabs-motion/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isVertical = exports.isRTL = exports.isIntrinsic = exports.isHorizontal = exports.isFixed = exports.getTabPanelId = exports.getTabId = void 0;

var _constants = require("./constants");

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
var getTabId = function getTabId(uid, key) {
  return "tabs-".concat(uid, "-tab-").concat(key);
};

exports.getTabId = getTabId;

var getTabPanelId = function getTabPanelId(uid, key) {
  return "tabs-".concat(uid, "-tabpanel-").concat(key);
};

exports.getTabPanelId = getTabPanelId;

var isHorizontal = function isHorizontal(orientation) {
  return orientation === _constants.ORIENTATION.horizontal;
};

exports.isHorizontal = isHorizontal;

var isVertical = function isVertical(orientation) {
  return orientation === _constants.ORIENTATION.vertical;
};

exports.isVertical = isVertical;

var isRTL = function isRTL(direction) {
  return direction === 'rtl';
};

exports.isRTL = isRTL;

var isIntrinsic = function isIntrinsic(fill) {
  return fill === _constants.FILL.intrinsic;
};

exports.isIntrinsic = isIntrinsic;

var isFixed = function isFixed(fill) {
  return fill === _constants.FILL.fixed;
};

exports.isFixed = isFixed;
