// baseui/themes/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  LightTheme: true,
  LightThemeMove: true,
  DarkTheme: true,
  darkThemeOverrides: true,
  DarkThemeMove: true,
  createDarkTheme: true,
  createLightTheme: true,
  createTheme: true,
  darkThemePrimitives: true,
  lightThemePrimitives: true
};
Object.defineProperty(exports, "DarkTheme", {
  enumerable: true,
  get: function get() {
    return _darkTheme.DarkTheme;
  }
});
Object.defineProperty(exports, "DarkThemeMove", {
  enumerable: true,
  get: function get() {
    return _darkThemeWithMove.DarkThemeMove;
  }
});
Object.defineProperty(exports, "LightTheme", {
  enumerable: true,
  get: function get() {
    return _lightTheme.LightTheme;
  }
});
Object.defineProperty(exports, "LightThemeMove", {
  enumerable: true,
  get: function get() {
    return _lightThemeWithMove.LightThemeMove;
  }
});
Object.defineProperty(exports, "createDarkTheme", {
  enumerable: true,
  get: function get() {
    return _createDarkTheme.default;
  }
});
Object.defineProperty(exports, "createLightTheme", {
  enumerable: true,
  get: function get() {
    return _createLightTheme.default;
  }
});
Object.defineProperty(exports, "createTheme", {
  enumerable: true,
  get: function get() {
    return _createLightTheme.default;
  }
});
Object.defineProperty(exports, "darkThemeOverrides", {
  enumerable: true,
  get: function get() {
    return _darkTheme.DarkTheme;
  }
});
Object.defineProperty(exports, "darkThemePrimitives", {
  enumerable: true,
  get: function get() {
    return _primitives.default;
  }
});
Object.defineProperty(exports, "lightThemePrimitives", {
  enumerable: true,
  get: function get() {
    return _primitives2.default;
  }
});

var _lightTheme = require("./light-theme/light-theme");

var _lightThemeWithMove = require("./move-theme/light-theme-with-move");

var _darkTheme = require("./dark-theme/dark-theme");

var _darkThemeWithMove = require("./move-theme/dark-theme-with-move");

var _createDarkTheme = _interopRequireDefault(require("./dark-theme/create-dark-theme"));

var _createLightTheme = _interopRequireDefault(require("./light-theme/create-light-theme"));

var _primitives = _interopRequireDefault(require("./dark-theme/primitives"));

var _primitives2 = _interopRequireDefault(require("./light-theme/primitives"));

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
