// baseui/map-marker/floating-marker.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _styles = require("../styles");

var _pinHead = _interopRequireDefault(require("./pin-head"));

var _overrides = require("../helpers/overrides");

var _styledComponents = require("./styled-components");

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function getAnchorPinHeadSize(anchorType) {
  if (anchorType === _constants.FLOATING_MARKER_ANCHOR_TYPES.circle) {
    return _constants.PINHEAD_SIZES_SHAPES.xSmallCircle;
  } else if (anchorType === _constants.FLOATING_MARKER_ANCHOR_TYPES.square) {
    return _constants.PINHEAD_SIZES_SHAPES.xSmallSquare;
  } else if (anchorType === _constants.FLOATING_MARKER_ANCHOR_TYPES.xxSmallSquare) {
    return _constants.PINHEAD_SIZES_SHAPES.xxSmallSquare;
  } else {
    return _constants.PINHEAD_SIZES_SHAPES.xxSmallCircle;
  }
}

var FloatingMarker = function FloatingMarker(_ref) {
  var label = _ref.label,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? _constants.PINHEAD_SIZES_SHAPES.medium : _ref$size,
      _ref$anchor = _ref.anchor,
      anchor = _ref$anchor === void 0 ? _constants.FLOATING_MARKER_ANCHOR_POSITIONS.bottomLeft : _ref$anchor,
      endEnhancer = _ref.endEnhancer,
      startEnhancer = _ref.startEnhancer,
      _ref$anchorType = _ref.anchorType,
      anchorType = _ref$anchorType === void 0 ? _constants.FLOATING_MARKER_ANCHOR_TYPES.circle : _ref$anchorType,
      _ref$overrides = _ref.overrides,
      overrides = _ref$overrides === void 0 ? {} : _ref$overrides;

  var _useStyletron = (0, _styles.useStyletron)(),
      _useStyletron2 = _slicedToArray(_useStyletron, 2),
      theme = _useStyletron2[1];

  var _theme$colors = theme.colors,
      backgroundPrimary = _theme$colors.backgroundPrimary,
      backgroundInversePrimary = _theme$colors.backgroundInversePrimary,
      primaryA = _theme$colors.primaryA,
      primaryB = _theme$colors.primaryB;
  var anchorPinHeadSize = getAnchorPinHeadSize(anchorType);

  var _getOverrides = (0, _overrides.getOverrides)(overrides.Root, _styledComponents.StyledFloatingMarkerRoot),
      _getOverrides2 = _slicedToArray(_getOverrides, 2),
      Root = _getOverrides2[0],
      rootProps = _getOverrides2[1];

  var _getOverrides3 = (0, _overrides.getOverrides)(overrides.PinHeadContainer, _styledComponents.StyledFloatingMarkerPinHeadContainer),
      _getOverrides4 = _slicedToArray(_getOverrides3, 2),
      FloatingMarkerPinHeadContainer = _getOverrides4[0],
      floatingMarkerPinHeadContainerProps = _getOverrides4[1];

  var _getOverrides5 = (0, _overrides.getOverrides)(overrides.AnchorContainer, _styledComponents.StyledFloatingMarkerAnchorContainer),
      _getOverrides6 = _slicedToArray(_getOverrides5, 2),
      FloatingMarkerAnchorContainer = _getOverrides6[0],
      floatingMarkerAnchorContainerProps = _getOverrides6[1];

  return /*#__PURE__*/React.createElement(Root, _extends({
    "data-baseweb": "floating-map-marker",
    $size: _constants.PINHEAD_DIMENSIONS[anchorPinHeadSize].height
  }, rootProps), /*#__PURE__*/React.createElement(FloatingMarkerPinHeadContainer, _extends({
    $anchor: anchor,
    $anchorSize: _constants.PINHEAD_DIMENSIONS[anchorPinHeadSize].height
  }, floatingMarkerPinHeadContainerProps), /*#__PURE__*/React.createElement(_pinHead.default, {
    size: size,
    color: primaryA,
    background: backgroundPrimary,
    type: _constants.PINHEAD_TYPES.floating,
    label: label,
    startEnhancer: startEnhancer,
    endEnhancer: endEnhancer,
    overrides: overrides
  })), anchor !== _constants.FLOATING_MARKER_ANCHOR_POSITIONS.none && /*#__PURE__*/React.createElement(FloatingMarkerAnchorContainer, floatingMarkerAnchorContainerProps, /*#__PURE__*/React.createElement(_pinHead.default, {
    size: anchorPinHeadSize,
    color: primaryB,
    background: backgroundInversePrimary,
    type: _constants.PINHEAD_TYPES.fixed,
    overrides: overrides
  })));
};

var _default = FloatingMarker;
exports.default = _default;
