// baseui/esm/helper/stateful-helper.js
var _excluded = ["children"];

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import * as React from 'react';
import { StatefulContainer, ACCESSIBILITY_TYPE, PLACEMENT, TRIGGER_TYPE } from '../popover';
import { POPOVER_MARGIN } from '../popover/constants';
import { Helper } from './helper';
export function StatefulHelper(props) {
  var children = props.children,
      restProps = _objectWithoutProperties(props, _excluded);

  return /*#__PURE__*/React.createElement(StatefulContainer, restProps, function (helperProps) {
    return /*#__PURE__*/React.createElement(Helper, helperProps, children);
  });
}
StatefulHelper.defaultProps = {
  accessibilityType: ACCESSIBILITY_TYPE.menu,
  ignoreBoundary: false,
  overrides: {},
  onMouseEnterDelay: 200,
  onMouseLeaveDelay: 200,
  placement: PLACEMENT.bottom,
  showArrow: true,
  triggerType: TRIGGER_TYPE.click,
  dismissOnClickOutside: true,
  dismissOnEsc: true,
  stateReducer: function stateReducer(_, nextState) {
    return nextState;
  },
  popoverMargin: POPOVER_MARGIN
};
