// baseui/es/helper/helper.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import * as React from 'react';
import { Popover, PLACEMENT } from '../popover';
import { mergeOverrides } from '../helpers/overrides';
import { StyledArrow, StyledBody } from './styled-components';
export function Helper(props) {
  const {
    overrides = {},
    placement,
    ...restProps
  } = props;
  const mergedOverrides = mergeOverrides({
    Arrow: placement !== PLACEMENT.auto ? StyledArrow : {},
    Body: placement !== PLACEMENT.auto ? StyledBody : {},
    Inner: {
      style: ({
        $theme
      }) => {
        return {
          backgroundColor: $theme.colors.backgroundPrimary
        };
      }
    }
  }, overrides);
  return /*#__PURE__*/React.createElement(Popover, _extends({
    "data-baseweb": "helper"
  }, restProps, {
    autoFocus: false,
    placement: placement,
    overrides: mergedOverrides
  }));
}
Helper.defaultProps = { ...Popover.defaultProps,
  placement: PLACEMENT.bottom,
  showArrow: true
};
