// baseui/esm/button-timed/styled-components.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import { styled, withStyle } from '../styles/index';
import { StyledBaseButton } from '../button';
import { hexToRgb as hexToRgba } from '../styles/util';
export var BaseButtonTimed = withStyle(StyledBaseButton, function (_ref) {
  var $theme = _ref.$theme,
      $initialTime = _ref.$initialTime,
      $timeElapsed = _ref.$timeElapsed;
  var completionPercentage = $timeElapsed / $initialTime * 100;
  var timeLeft = $initialTime - $timeElapsed;
  return _objectSpread({
    position: 'relative'
  }, $initialTime > 0 ? {
    ':after': {
      animationDuration: "".concat(timeLeft, "s"),
      animationName: {
        from: {
          transform: "translateX(".concat(completionPercentage, "%)")
        },
        to: {
          transform: 'translateX(100%)'
        }
      },
      animationTimingFunction: 'linear',
      display: 'inline-block',
      content: '""',
      width: '100%',
      height: '100%',
      zIndex: '1',
      position: 'absolute',
      backgroundColor: hexToRgba($theme.colors.backgroundPrimary, '0.2')
    }
  } : {});
});
BaseButtonTimed.displayName = "BaseButtonTimed";
export var TimerContainer = styled('span', {
  // minWidth to ensure button width stays consistent as timeRemaining changes
  minWidth: '53px'
});
TimerContainer.displayName = "TimerContainer";
