// baseui/textarea/styled-components.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledTextareaContainer = exports.StyledTextarea = exports.StyledTextAreaRoot = void 0;

var _styles = require("../styles");

var _styledComponents = require("../input/styled-components");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var StyledTextAreaRoot = (0, _styles.styled)('div', function (props) {
  return _objectSpread(_objectSpread({}, (0, _styledComponents.getRootStyles)(_objectSpread(_objectSpread({
    $positive: false
  }, props), {}, {
    $hasIconTrailing: false
  }))), {}, {
    width: props.$resize ? 'fit-content' : '100%'
  });
});
exports.StyledTextAreaRoot = StyledTextAreaRoot;
StyledTextAreaRoot.displayName = "StyledTextAreaRoot";
StyledTextAreaRoot.displayName = 'StyledTextAreaRoot';
var StyledTextareaContainer = (0, _styles.styled)('div', function (props) {
  return (0, _styledComponents.getInputContainerStyles)(_objectSpread({
    $positive: false
  }, props));
});
exports.StyledTextareaContainer = StyledTextareaContainer;
StyledTextareaContainer.displayName = "StyledTextareaContainer";
StyledTextareaContainer.displayName = 'StyledTextareaContainer';
var StyledTextarea = (0, _styles.styled)('textarea', function (props) {
  return _objectSpread(_objectSpread({}, (0, _styledComponents.getInputStyles)(props)), {}, {
    resize: props.$resize || 'none'
  });
});
exports.StyledTextarea = StyledTextarea;
StyledTextarea.displayName = "StyledTextarea";
StyledTextarea.displayName = 'StyledTextarea';
