// baseui/esm/pagination/stateful-pagination.js
var _excluded = ["numPages", "initialState", "stateReducer", "onPageChange"];

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import * as React from 'react';
import Pagination from './pagination';
import StatefulContainer from './stateful-container';
export default function StatefulPagination(_ref) {
  var numPages = _ref.numPages,
      initialState = _ref.initialState,
      stateReducer = _ref.stateReducer,
      onPageChange = _ref.onPageChange,
      props = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React.createElement(StatefulContainer, {
    numPages: numPages,
    initialState: initialState,
    stateReducer: stateReducer,
    onPageChange: onPageChange
  }, function (renderProps) {
    return /*#__PURE__*/React.createElement(Pagination, _extends({
      numPages: numPages
    }, renderProps, props));
  });
}
StatefulPagination.defaultProps = {
  initialState: {
    currentPage: 1
  },
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  stateReducer: function stateReducer(changeType, changes) {
    return changes;
  },
  overrides: {}
};
