// baseui/snackbar/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  DURATION: true,
  PLACEMENT: true,
  SnackbarElement: true,
  SnackbarProvider: true,
  useSnackbar: true,
  StyledRoot: true,
  StyledContent: true,
  StyledStartEnhancerContainer: true,
  StyledSpinner: true,
  StyledMessage: true,
  StyledWrapActionButtonContainer: true,
  StyledActionButtonContainer: true,
  StyledPlacementContainer: true
};
Object.defineProperty(exports, "DURATION", {
  enumerable: true,
  get: function get() {
    return _constants.DURATION;
  }
});
Object.defineProperty(exports, "PLACEMENT", {
  enumerable: true,
  get: function get() {
    return _constants.PLACEMENT;
  }
});
Object.defineProperty(exports, "SnackbarElement", {
  enumerable: true,
  get: function get() {
    return _snackbarElement.default;
  }
});
Object.defineProperty(exports, "SnackbarProvider", {
  enumerable: true,
  get: function get() {
    return _snackbarContext.default;
  }
});
Object.defineProperty(exports, "StyledActionButtonContainer", {
  enumerable: true,
  get: function get() {
    return _styledComponents.StyledActionButtonContainer;
  }
});
Object.defineProperty(exports, "StyledContent", {
  enumerable: true,
  get: function get() {
    return _styledComponents.StyledContent;
  }
});
Object.defineProperty(exports, "StyledMessage", {
  enumerable: true,
  get: function get() {
    return _styledComponents.StyledMessage;
  }
});
Object.defineProperty(exports, "StyledPlacementContainer", {
  enumerable: true,
  get: function get() {
    return _styledComponents.StyledPlacementContainer;
  }
});
Object.defineProperty(exports, "StyledRoot", {
  enumerable: true,
  get: function get() {
    return _styledComponents.StyledRoot;
  }
});
Object.defineProperty(exports, "StyledSpinner", {
  enumerable: true,
  get: function get() {
    return _styledComponents.StyledSpinner;
  }
});
Object.defineProperty(exports, "StyledStartEnhancerContainer", {
  enumerable: true,
  get: function get() {
    return _styledComponents.StyledStartEnhancerContainer;
  }
});
Object.defineProperty(exports, "StyledWrapActionButtonContainer", {
  enumerable: true,
  get: function get() {
    return _styledComponents.StyledWrapActionButtonContainer;
  }
});
Object.defineProperty(exports, "useSnackbar", {
  enumerable: true,
  get: function get() {
    return _snackbarContext.useSnackbar;
  }
});

var _constants = require("./constants");

var _snackbarElement = _interopRequireDefault(require("./snackbar-element"));

var _snackbarContext = _interopRequireWildcard(require("./snackbar-context"));

var _styledComponents = require("./styled-components");

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
