// baseui/themes/light-theme/color-tokens.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _tokens = require("../../tokens");

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
// color constants
var lightColorTokens = {
  // Primary Palette
  primaryA: _tokens.colors.black,
  primaryB: _tokens.colors.white,
  primary: _tokens.colors.black,
  primary50: _tokens.colors.gray50,
  primary100: _tokens.colors.gray100,
  primary200: _tokens.colors.gray200,
  primary300: _tokens.colors.gray300,
  primary400: _tokens.colors.gray400,
  primary500: _tokens.colors.gray500,
  primary600: _tokens.colors.gray600,
  primary700: _tokens.colors.gray700,
  // Accent Palette
  accent: _tokens.colors.blue400,
  accent50: _tokens.colors.blue50,
  accent100: _tokens.colors.blue100,
  accent200: _tokens.colors.blue200,
  accent300: _tokens.colors.blue300,
  accent400: _tokens.colors.blue400,
  accent500: _tokens.colors.blue500,
  accent600: _tokens.colors.blue600,
  accent700: _tokens.colors.blue700,
  // Negative Palette
  negative: _tokens.colors.red400,
  negative50: _tokens.colors.red50,
  negative100: _tokens.colors.red100,
  negative200: _tokens.colors.red200,
  negative300: _tokens.colors.red300,
  negative400: _tokens.colors.red400,
  negative500: _tokens.colors.red500,
  negative600: _tokens.colors.red600,
  negative700: _tokens.colors.red700,
  // Warning Palette
  warning: _tokens.colors.yellow400,
  warning50: _tokens.colors.yellow50,
  warning100: _tokens.colors.yellow100,
  warning200: _tokens.colors.yellow200,
  warning300: _tokens.colors.yellow300,
  warning400: _tokens.colors.yellow400,
  warning500: _tokens.colors.yellow500,
  warning600: _tokens.colors.yellow600,
  warning700: _tokens.colors.yellow700,
  // Positive Palette
  positive: _tokens.colors.green500,
  positive50: _tokens.colors.green50,
  positive100: _tokens.colors.green100,
  positive200: _tokens.colors.green200,
  positive300: _tokens.colors.green300,
  positive400: _tokens.colors.green400,
  positive500: _tokens.colors.green500,
  positive600: _tokens.colors.green600,
  positive700: _tokens.colors.green700,
  // Monochrome Palette
  white: _tokens.colors.white,
  black: _tokens.colors.black,
  mono100: _tokens.colors.white,
  mono200: _tokens.colors.gray50,
  mono300: _tokens.colors.gray100,
  mono400: _tokens.colors.gray200,
  mono500: _tokens.colors.gray300,
  mono600: _tokens.colors.gray400,
  mono700: _tokens.colors.gray500,
  mono800: _tokens.colors.gray600,
  mono900: _tokens.colors.gray700,
  mono1000: _tokens.colors.black,
  // Rating Palette,
  ratingInactiveFill: _tokens.colors.gray100,
  ratingStroke: _tokens.colors.gray300
};
var _default = lightColorTokens;
exports.default = _default;
