// baseui/banner/styled-components.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledTrailingIconButton = exports.StyledTrailingContent = exports.StyledTrailingButtonContainer = exports.StyledTitle = exports.StyledRoot = exports.StyledMessageContent = exports.StyledMessage = exports.StyledLeadingContent = exports.StyledBelowContent = void 0;

var _styles = require("../styles");

var _constants = require("./constants");

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
var StyledRoot = (0, _styles.styled)('div', function (_ref) {
  var $theme = _ref.$theme,
      $backgroundColor = _ref.$backgroundColor,
      $color = _ref.$color,
      $nested = _ref.$nested;
  var radius = $nested ? $theme.borders.radius300 : $theme.borders.radius400;
  return {
    backgroundColor: $backgroundColor,
    borderStartStartRadius: radius,
    borderStartEndRadius: radius,
    borderEndStartRadius: radius,
    borderEndEndRadius: radius,
    color: $color,
    display: 'grid',
    gridColumnGap: $theme.sizing.scale600,
    gridTemplateColumns: 'min-content auto min-content',
    gridTemplateRows: 'auto min-content',
    marginTop: $theme.sizing.scale600,
    marginRight: $theme.sizing.scale600,
    marginBottom: $theme.sizing.scale600,
    marginLeft: $theme.sizing.scale600
  };
});
exports.StyledRoot = StyledRoot;
StyledRoot.displayName = "StyledRoot";
StyledRoot.displayName = 'StyledRoot';
var StyledLeadingContent = (0, _styles.styled)('div', function (_ref2) {
  var $theme = _ref2.$theme,
      $includesArtwork = _ref2.$includesArtwork;
  return {
    alignItems: 'center',
    display: 'flex',
    paddingInlineStart: $includesArtwork ? $theme.sizing.scale600 : null
  };
});
exports.StyledLeadingContent = StyledLeadingContent;
StyledLeadingContent.displayName = "StyledLeadingContent";
StyledLeadingContent.displayName = 'StyledLeadingContent';
var StyledMessageContent = (0, _styles.styled)('div', function (_ref3) {
  var $theme = _ref3.$theme,
      $actionPosition = _ref3.$actionPosition;
  return {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    paddingBlockStart: $theme.sizing.scale600,
    paddingBlockEnd: $actionPosition === _constants.ACTION_POSITION.trailing ? $theme.sizing.scale600 : null
  };
});
exports.StyledMessageContent = StyledMessageContent;
StyledMessageContent.displayName = "StyledMessageContent";
StyledMessageContent.displayName = 'StyledMessageContent'; // eslint-disable-next-line @typescript-eslint/no-unused-vars

var StyledTrailingContent = (0, _styles.styled)('div', function (_ref4) {
  var $theme = _ref4.$theme;
  return {
    display: 'flex',
    gridRowEnd: 'span 2',
    marginInlineStart: 'auto'
  };
});
exports.StyledTrailingContent = StyledTrailingContent;
StyledTrailingContent.displayName = "StyledTrailingContent";
StyledTrailingContent.displayName = 'StyledTrailingContent';
var StyledBelowContent = (0, _styles.styled)('div', function (_ref5) {
  var $theme = _ref5.$theme,
      $actionPosition = _ref5.$actionPosition;
  return {
    gridColumnStart: 2,
    paddingBlockStart: $actionPosition === _constants.ACTION_POSITION.below ? $theme.sizing.scale300 : null,
    paddingBlockEnd: $actionPosition === _constants.ACTION_POSITION.below ? $theme.sizing.scale600 : null
  };
});
exports.StyledBelowContent = StyledBelowContent;
StyledBelowContent.displayName = "StyledBelowContent";
StyledBelowContent.displayName = 'StyledBelowContent';
var StyledTitle = (0, _styles.styled)('div', function (_ref6) {
  var $theme = _ref6.$theme;
  return $theme.typography.LabelMedium;
});
exports.StyledTitle = StyledTitle;
StyledTitle.displayName = "StyledTitle";
StyledTitle.displayName = 'StyledTitle';
var StyledMessage = (0, _styles.styled)('div', function (_ref7) {
  var $theme = _ref7.$theme;
  return $theme.typography.ParagraphMedium;
});
exports.StyledMessage = StyledMessage;
StyledMessage.displayName = "StyledMessage";
StyledMessage.displayName = 'StyledMessage';
var StyledTrailingButtonContainer = (0, _styles.styled)('div', function (_ref8) {
  var $theme = _ref8.$theme;
  return {
    display: 'flex',
    alignItems: 'center',
    paddingInlineEnd: $theme.sizing.scale600
  };
});
exports.StyledTrailingButtonContainer = StyledTrailingButtonContainer;
StyledTrailingButtonContainer.displayName = "StyledTrailingButtonContainer";
StyledTrailingButtonContainer.displayName = 'StyledTrailingButtonContainer';
var StyledTrailingIconButton = (0, _styles.styled)('button', function (_ref9) {
  var $theme = _ref9.$theme,
      $nested = _ref9.$nested;
  var radius = $nested ? $theme.borders.radius300 : $theme.borders.radius400;
  return {
    alignItems: 'center',
    background: 'none',
    border: 'none',
    borderStartEndRadius: radius,
    borderEndEndRadius: radius,
    color: 'inherit',
    cursor: 'pointer',
    display: 'flex',
    paddingInlineStart: $theme.sizing.scale600,
    paddingInlineEnd: $theme.sizing.scale600,
    ':hover': {
      boxShadow: 'inset 999px 999px 0px rgba(0, 0, 0, 0.04)'
    },
    ':active': {
      boxShadow: 'inset 999px 999px 0px rgba(0, 0, 0, 0.08)'
    }
  };
});
exports.StyledTrailingIconButton = StyledTrailingIconButton;
StyledTrailingIconButton.displayName = "StyledTrailingIconButton";
StyledTrailingIconButton.displayName = 'StyledTrailingIconButton';
