// baseui/esm/message-card/message-card.js
var _excluded = ["children"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import * as React from 'react';
import { StyledRoot, StyledImage, StyledContentContainer, StyledHeadingContainer, StyledParagraphContainer } from './styled-components';
import { Button as DefaultButton, SHAPE, SIZE } from '../button';
import { useStyletron } from '../styles/index';
import { ThemeProvider, LightTheme } from '../';
import { getBackgroundColorType } from './utils';
import { colors } from '../tokens';
import { getOverrides } from '../helpers/overrides';
import { IMAGE_LAYOUT, BACKGROUND_COLOR_TYPE, BUTTON_KIND } from './constants';

var ButtonAlwaysLightTheme = function ButtonAlwaysLightTheme(_ref) {
  var children = _ref.children,
      restProps = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React.createElement(ThemeProvider, {
    theme: LightTheme
  }, /*#__PURE__*/React.createElement(DefaultButton, restProps, children));
};

var MessageCard = function MessageCard(_ref2) {
  var _ref2$backgroundColor = _ref2.backgroundColor,
      backgroundColor = _ref2$backgroundColor === void 0 ? colors.white : _ref2$backgroundColor,
      backgroundColorTypeProp = _ref2.backgroundColorType,
      _ref2$buttonKind = _ref2.buttonKind,
      buttonKind = _ref2$buttonKind === void 0 ? BUTTON_KIND.secondary : _ref2$buttonKind,
      buttonLabel = _ref2.buttonLabel,
      heading = _ref2.heading,
      image = _ref2.image,
      onClick = _ref2.onClick,
      _ref2$overrides = _ref2.overrides,
      overrides = _ref2$overrides === void 0 ? {} : _ref2$overrides,
      paragraph = _ref2.paragraph;

  var _ref3 = image || {},
      src = _ref3.src,
      _ref3$layout = _ref3.layout,
      layout = _ref3$layout === void 0 ? IMAGE_LAYOUT.top : _ref3$layout,
      backgroundPosition = _ref3.backgroundPosition,
      ariaLabel = _ref3.ariaLabel;

  var _getOverrides = getOverrides(overrides.Root, StyledRoot),
      _getOverrides2 = _slicedToArray(_getOverrides, 2),
      Root = _getOverrides2[0],
      RootProps = _getOverrides2[1];

  var _getOverrides3 = getOverrides(overrides.ContentContainer, StyledContentContainer),
      _getOverrides4 = _slicedToArray(_getOverrides3, 2),
      ContentContainer = _getOverrides4[0],
      ContentContainerProps = _getOverrides4[1];

  var _getOverrides5 = getOverrides(overrides.HeadingContainer, StyledHeadingContainer),
      _getOverrides6 = _slicedToArray(_getOverrides5, 2),
      HeadingContainer = _getOverrides6[0],
      HeadingContainerProps = _getOverrides6[1];

  var _getOverrides7 = getOverrides(overrides.ParagraphContainer, StyledParagraphContainer),
      _getOverrides8 = _slicedToArray(_getOverrides7, 2),
      ParagraphContainer = _getOverrides8[0],
      ParagraphContainerProps = _getOverrides8[1];

  var _getOverrides9 = getOverrides(overrides.Image, StyledImage),
      _getOverrides10 = _slicedToArray(_getOverrides9, 2),
      Image = _getOverrides10[0],
      ImageProps = _getOverrides10[1];

  var _getOverrides11 = getOverrides(overrides.Button, ButtonAlwaysLightTheme),
      _getOverrides12 = _slicedToArray(_getOverrides11, 2),
      Button = _getOverrides12[0],
      ButtonProps = _getOverrides12[1];

  var _useStyletron = useStyletron(),
      _useStyletron2 = _slicedToArray(_useStyletron, 2),
      theme = _useStyletron2[1];

  var backgroundColorType = backgroundColorTypeProp || getBackgroundColorType(backgroundColor);

  if (!backgroundColorType) {
    backgroundColorType = BACKGROUND_COLOR_TYPE.dark;

    if (process.env.NODE_ENV !== "production") {
      console.warn("The provided value for 'backgroundColor', ".concat(backgroundColor, ", is not recognized as a         Base Web primitive color. Please use the 'backgroundColorType' prop to indicate whether         this color is light or dark so the rest of the component can be styled accordingly."));
    }
  }

  var buttonColors;

  if (buttonKind === BUTTON_KIND.tertiary && backgroundColorType === BACKGROUND_COLOR_TYPE.dark) {
    buttonColors = {
      color: theme.colors.contentOnColor,
      backgroundColor: theme.colors.buttonTertiaryFill
    };
  }

  if (buttonKind !== BUTTON_KIND.tertiary && backgroundColor !== colors.white) {
    buttonColors = {
      color: theme.colors.contentOnColorInverse,
      backgroundColor: theme.colors.backgroundAlwaysLight
    };
  }

  return /*#__PURE__*/React.createElement(Root, _extends({
    onClick: onClick,
    $backgroundColor: backgroundColor,
    $backgroundColorType: backgroundColorType,
    $imageLayout: layout
  }, RootProps), image && /*#__PURE__*/React.createElement(Image, _extends({
    role: "img",
    "aria-label": ariaLabel,
    $src: src,
    $imageLayout: layout,
    $backgroundPosition: backgroundPosition
  }, ImageProps)), /*#__PURE__*/React.createElement(ContentContainer, ContentContainerProps, heading && /*#__PURE__*/React.createElement(HeadingContainer, HeadingContainerProps, heading), paragraph && /*#__PURE__*/React.createElement(ParagraphContainer, ParagraphContainerProps, paragraph), buttonLabel && /*#__PURE__*/React.createElement(Button, _extends({
    $as: "div",
    kind: buttonKind,
    shape: SHAPE.pill,
    size: SIZE.compact,
    role: "none",
    tabIndex: -1,
    colors: buttonColors,
    overrides: {
      BaseButton: {
        style: _objectSpread({
          textAlign: 'center',
          pointerEvents: 'none'
        }, buttonKind === BUTTON_KIND.tertiary ? {
          marginTop: theme.sizing.scale100,
          transform: theme.direction === 'rtl' ? "translateX(".concat(theme.sizing.scale500, ")") : "translateX(-".concat(theme.sizing.scale500, ")")
        } : {
          marginTop: theme.sizing.scale500
        })
      }
    }
  }, ButtonProps), buttonLabel)));
};

export default MessageCard;
