// baseui/es/toast/toast.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/

/* global document */
import * as React from 'react';
import { getOverrides, mergeOverrides } from '../helpers/overrides';
import DeleteIcon from '../icon/delete';
import { Body as StyledBody, CloseIconSvg as StyledCloseIcon, InnerContainer as StyledInnerContainer } from './styled-components';
import { KIND, TYPE } from './constants';
import { LocaleContext } from '../locale';
import { isFocusVisible, forkFocus, forkBlur } from '../utils/focusVisible';

class Toast extends React.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "autoHideTimeout", void 0);

    _defineProperty(this, "animateInTimer", void 0);

    _defineProperty(this, "animateOutCompleteTimer", void 0);

    _defineProperty(this, "closeRef", void 0);

    _defineProperty(this, "previouslyFocusedElement", void 0);

    _defineProperty(this, "state", {
      isVisible: false,
      isRendered: true,
      isFocusVisible: false
    });

    _defineProperty(this, "handleFocus", event => {
      if (isFocusVisible(event)) {
        this.setState({
          isFocusVisible: true
        });
      }
    });

    _defineProperty(this, "handleBlur", event => {
      if (this.state.isFocusVisible !== false) {
        this.setState({
          isFocusVisible: false
        });
      }
    });

    _defineProperty(this, "animateIn", () => {
      // Defer to next event loop
      this.animateInTimer = setTimeout(() => {
        this.setState({
          isVisible: true
        });
      }, 0);
    });

    _defineProperty(this, "animateOut", (callback = () => {}) => {
      this.setState({
        isVisible: false
      }); // Remove the toast from the DOM after animation finishes

      this.animateOutCompleteTimer = setTimeout(() => {
        this.setState({
          isRendered: false
        });
        callback();
      }, 600);
    });

    _defineProperty(this, "dismiss", () => {
      this.animateOut(this.props.onClose);

      if (this.props.autoFocus && this.previouslyFocusedElement) {
        this.previouslyFocusedElement.focus();
      }
    });

    _defineProperty(this, "onFocus", e => {
      if (!this.state.isVisible) return;
      clearTimeout(this.autoHideTimeout);
      clearTimeout(this.animateOutCompleteTimer);
      typeof this.props.onFocus === 'function' && this.props.onFocus(e);
    });

    _defineProperty(this, "onMouseEnter", e => {
      if (!this.state.isVisible) return;
      clearTimeout(this.autoHideTimeout);
      clearTimeout(this.animateOutCompleteTimer);
      typeof this.props.onMouseEnter === 'function' && this.props.onMouseEnter(e);
    });

    _defineProperty(this, "onBlur", e => {
      this.startTimeout();
      typeof this.props.onBlur === 'function' && this.props.onBlur(e);
    });

    _defineProperty(this, "onMouseLeave", e => {
      this.startTimeout();
      typeof this.props.onMouseLeave === 'function' && this.props.onMouseLeave(e);
    });

    this.closeRef = /*#__PURE__*/React.createRef();
    this.previouslyFocusedElement = null;
  }

  componentDidMount() {
    this.animateIn();
    this.startTimeout();

    if (typeof document !== 'undefined' && this.props.autoFocus && this.closeRef && this.closeRef.current && this.closeRef.current.focus && typeof this.closeRef.current.focus === 'function') {
      // todo(flow->ts): double check if typecast is correct
      this.previouslyFocusedElement = document.activeElement;
      this.closeRef.current.focus();
      this.setState({
        isFocusVisible: true
      });
    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.autoHideDuration !== prevProps.autoHideDuration || this.props.__updated !== prevProps.__updated) {
      this.startTimeout();
    }
  }

  componentWillUnmount() {
    this.clearTimeout();
  }

  startTimeout() {
    if (this.props.autoHideDuration) {
      if (this.autoHideTimeout) {
        clearTimeout(this.autoHideTimeout);
      }

      this.autoHideTimeout = setTimeout(this.dismiss, this.props.autoHideDuration);
    }
  }

  clearTimeout() {
    [this.autoHideTimeout, this.animateInTimer, this.animateOutCompleteTimer].forEach(timerId => {
      if (timerId) {
        clearTimeout(timerId);
      }
    });
  }

  getSharedProps() {
    const {
      kind,
      notificationType,
      closeable
    } = this.props;
    const {
      isRendered,
      isVisible
    } = this.state;
    return {
      $kind: kind,
      $type: notificationType,
      $closeable: closeable,
      $isRendered: isRendered,
      $isVisible: isVisible
    };
  }

  render() {
    const {
      children,
      closeable
    } = this.props;
    const {
      isRendered
    } = this.state;
    const {
      Body: BodyOverride,
      CloseIcon: CloseIconOverride,
      InnerContainer: InnerContainerOverride
    } = this.props.overrides;
    const [Body, bodyProps] = getOverrides(BodyOverride, StyledBody);
    const [InnerContainer, innerContainerProps] = getOverrides(InnerContainerOverride, StyledInnerContainer);
    const [CloseIcon, closeIconProps] = getOverrides(CloseIconOverride, StyledCloseIcon);
    const closeIconOverrides = mergeOverrides({
      Svg: {
        component: CloseIcon
      }
    }, {
      Svg: CloseIconOverride
    });
    const sharedProps = this.getSharedProps();

    if (!isRendered) {
      return null;
    }

    return /*#__PURE__*/React.createElement(LocaleContext.Consumer, null, locale => /*#__PURE__*/React.createElement(Body, _extends({
      role: "alert",
      "data-baseweb": this.props['data-baseweb'] || 'toast'
    }, sharedProps, bodyProps, {
      // the properties below have to go after overrides
      onBlur: this.onBlur,
      onFocus: this.onFocus,
      onMouseEnter: this.onMouseEnter,
      onMouseLeave: this.onMouseLeave
    }), /*#__PURE__*/React.createElement(InnerContainer, _extends({}, sharedProps, innerContainerProps), typeof children === 'function' ? children({
      dismiss: this.dismiss
    }) : children), closeable ? /*#__PURE__*/React.createElement(DeleteIcon, _extends({
      ref: this.closeRef,
      role: "button",
      tabIndex: 0,
      $isFocusVisible: this.state.isFocusVisible,
      onClick: this.dismiss,
      onKeyPress: event => {
        if (event.key === 'Enter') {
          this.dismiss();
        }
      },
      title: locale.toast.close
    }, sharedProps, closeIconProps, {
      onFocus: forkFocus(closeIconProps, this.handleFocus),
      onBlur: forkBlur(closeIconProps, this.handleBlur),
      overrides: closeIconOverrides
    })) : null));
  }

}

_defineProperty(Toast, "defaultProps", {
  autoFocus: false,
  autoHideDuration: 0,
  closeable: true,
  kind: KIND.info,
  notificationType: TYPE.toast,
  // Do we need a separate handler for
  // when a notification dismisses automatically
  onClose: () => {},
  onBlur: () => {},
  onFocus: () => {},
  onMouseEnter: () => {},
  onMouseLeave: () => {},
  overrides: {}
});

export default Toast;
