// baseui/esm/map-marker/location-puck.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import * as React from 'react';
import { useStyletron } from '../styles';
import { getOverrides } from '../helpers/overrides';
import { LOCATION_PUCK_SIZES, LOCATION_PUCK_TYPES, EARNER_LOCATION_PUCK_CORE_SCALES } from './constants';
import { LocationPuckContainer, StyledConsumerLocationPuckCore, StyledLocationPuckApproximation, StyledEarnerLocationPuckCore } from './styled-components';

var ConsumerLocationPuckHeading = function ConsumerLocationPuckHeading(_ref) {
  var heading = _ref.heading;

  var _useStyletron = useStyletron(),
      _useStyletron2 = _slicedToArray(_useStyletron, 2),
      css = _useStyletron2[0],
      theme = _useStyletron2[1];

  return /*#__PURE__*/React.createElement("svg", {
    className: css({
      height: "".concat(6, "px"),
      width: "".concat(11, "px"),
      position: 'absolute',
      color: theme.colors.contentAccent,
      transition: "".concat(theme.animation.timing300, " ").concat(theme.animation.easeOutCurve, " all")
    }),
    style: {
      transform: "rotate(".concat(heading, "deg) translateY(-16px)")
    },
    width: "11",
    height: "6",
    viewBox: "0 0 11 6",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M5.5 0L0 6L5.5 5L11 6L5.5 0Z",
    fill: "currentColor"
  }));
};

var ConsumerLocationPuck = function ConsumerLocationPuck(_ref2) {
  var heading = _ref2.heading,
      showHeading = _ref2.showHeading,
      confidenceRadius = _ref2.confidenceRadius,
      overrides = _ref2.overrides;

  var _useStyletron3 = useStyletron(),
      _useStyletron4 = _slicedToArray(_useStyletron3, 2),
      theme = _useStyletron4[1];

  var _getOverrides = getOverrides(overrides.Root, LocationPuckContainer),
      _getOverrides2 = _slicedToArray(_getOverrides, 2),
      Root = _getOverrides2[0],
      rootProps = _getOverrides2[1];

  var _getOverrides3 = getOverrides(overrides.ConsumerLocationPuckCore, StyledConsumerLocationPuckCore),
      _getOverrides4 = _slicedToArray(_getOverrides3, 2),
      ConsumerLocationPuckCore = _getOverrides4[0],
      consumerLocationPuckCoreProps = _getOverrides4[1];

  var _getOverrides5 = getOverrides(overrides.LocationPuckApproximation, StyledLocationPuckApproximation),
      _getOverrides6 = _slicedToArray(_getOverrides5, 2),
      LocationPuckApproximation = _getOverrides6[0],
      locationPuckApproximationProps = _getOverrides6[1];

  return /*#__PURE__*/React.createElement(Root, rootProps, /*#__PURE__*/React.createElement(LocationPuckApproximation, _extends({
    $color: theme.colors.contentAccent,
    $radius: confidenceRadius
  }, locationPuckApproximationProps)), /*#__PURE__*/React.createElement(ConsumerLocationPuckCore, consumerLocationPuckCoreProps), showHeading && /*#__PURE__*/React.createElement(ConsumerLocationPuckHeading, {
    heading: heading
  }));
};

var EarnerLocationPuckHeading = function EarnerLocationPuckHeading(_ref3) {
  var size = _ref3.size,
      color = _ref3.color,
      heading = _ref3.heading;

  var _useStyletron5 = useStyletron(),
      _useStyletron6 = _slicedToArray(_useStyletron5, 2),
      css = _useStyletron6[0],
      theme = _useStyletron6[1];

  return /*#__PURE__*/React.createElement("svg", {
    className: css({
      position: 'absolute',
      transition: "".concat(theme.animation.timing300, " ").concat(theme.animation.easeOutCurve, " all")
    }),
    style: {
      transform: "rotate(".concat(heading, "deg) scale(").concat(EARNER_LOCATION_PUCK_CORE_SCALES[size], ")")
    },
    width: "72",
    height: "72",
    viewBox: "0 0 72 72",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M36 20L52 52L36 42.8571L20 52L36 20Z",
    fill: color
  }));
};

var EarnerLocationPuck = function EarnerLocationPuck(_ref4) {
  var heading = _ref4.heading,
      confidenceRadius = _ref4.confidenceRadius,
      size = _ref4.size,
      overrides = _ref4.overrides;

  var _useStyletron7 = useStyletron(),
      _useStyletron8 = _slicedToArray(_useStyletron7, 2),
      theme = _useStyletron8[1];

  var color = theme.colors.contentPrimary;

  var _getOverrides7 = getOverrides(overrides.Root, LocationPuckContainer),
      _getOverrides8 = _slicedToArray(_getOverrides7, 2),
      Root = _getOverrides8[0],
      rootProps = _getOverrides8[1];

  var _getOverrides9 = getOverrides(overrides.LocationPuckApproximation, StyledLocationPuckApproximation),
      _getOverrides10 = _slicedToArray(_getOverrides9, 2),
      LocationPuckApproximation = _getOverrides10[0],
      locationPuckApproximationProps = _getOverrides10[1];

  var _getOverrides11 = getOverrides(overrides.EarnerLocationPuckCore, StyledEarnerLocationPuckCore),
      _getOverrides12 = _slicedToArray(_getOverrides11, 2),
      EarnerLocationPuckCore = _getOverrides12[0],
      earnerLocationPuckCoreProps = _getOverrides12[1];

  return /*#__PURE__*/React.createElement(Root, rootProps, /*#__PURE__*/React.createElement(LocationPuckApproximation, _extends({
    $color: color,
    $radius: confidenceRadius
  }, locationPuckApproximationProps)), /*#__PURE__*/React.createElement(EarnerLocationPuckCore, _extends({
    $color: color,
    $size: size
  }, earnerLocationPuckCoreProps)), /*#__PURE__*/React.createElement(EarnerLocationPuckHeading, {
    size: size,
    color: color,
    heading: heading
  }));
};

var LocationPuck = function LocationPuck(_ref5) {
  var _ref5$size = _ref5.size,
      size = _ref5$size === void 0 ? LOCATION_PUCK_SIZES.medium : _ref5$size,
      _ref5$heading = _ref5.heading,
      heading = _ref5$heading === void 0 ? 0 : _ref5$heading,
      _ref5$showHeading = _ref5.showHeading,
      showHeading = _ref5$showHeading === void 0 ? true : _ref5$showHeading,
      _ref5$confidenceRadiu = _ref5.confidenceRadius,
      confidenceRadius = _ref5$confidenceRadiu === void 0 ? 0 : _ref5$confidenceRadiu,
      _ref5$type = _ref5.type,
      type = _ref5$type === void 0 ? LOCATION_PUCK_TYPES.consumer : _ref5$type,
      _ref5$overrides = _ref5.overrides,
      overrides = _ref5$overrides === void 0 ? {} : _ref5$overrides;

  if (process.env.NODE_ENV !== "production") {
    if (size !== LOCATION_PUCK_SIZES.medium && type === LOCATION_PUCK_TYPES.consumer) {
      console.warn("Location puck size can only be applied to type === LOCATION_PUCK_TYPES.earner");
    }

    if (!showHeading && type === LOCATION_PUCK_TYPES.earner) {
      console.warn("Earner location puck must show the heading indicator");
    }
  }

  var sharedProps = {
    heading: heading,
    confidenceRadius: confidenceRadius,
    overrides: overrides
  };
  return type === LOCATION_PUCK_TYPES.consumer ? /*#__PURE__*/React.createElement(ConsumerLocationPuck, _extends({}, sharedProps, {
    showHeading: showHeading
  })) : /*#__PURE__*/React.createElement(EarnerLocationPuck, _extends({}, sharedProps, {
    size: size
  }));
};

export default LocationPuck;
