// baseui/toast/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  toaster: true,
  ToasterContainer: true,
  Toast: true,
  KIND: true,
  PLACEMENT: true,
  TYPE: true,
  StyledRoot: true,
  StyledBody: true,
  StyledCloseIcon: true
};
Object.defineProperty(exports, "KIND", {
  enumerable: true,
  get: function get() {
    return _constants.KIND;
  }
});
Object.defineProperty(exports, "PLACEMENT", {
  enumerable: true,
  get: function get() {
    return _constants.PLACEMENT;
  }
});
Object.defineProperty(exports, "StyledBody", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Body;
  }
});
Object.defineProperty(exports, "StyledCloseIcon", {
  enumerable: true,
  get: function get() {
    return _styledComponents.CloseIconSvg;
  }
});
Object.defineProperty(exports, "StyledRoot", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Root;
  }
});
Object.defineProperty(exports, "TYPE", {
  enumerable: true,
  get: function get() {
    return _constants.TYPE;
  }
});
Object.defineProperty(exports, "Toast", {
  enumerable: true,
  get: function get() {
    return _toast.default;
  }
});
Object.defineProperty(exports, "ToasterContainer", {
  enumerable: true,
  get: function get() {
    return _toaster.ToasterContainer;
  }
});
Object.defineProperty(exports, "toaster", {
  enumerable: true,
  get: function get() {
    return _toaster.default;
  }
});

var _toaster = _interopRequireWildcard(require("./toaster"));

var _toast = _interopRequireDefault(require("./toast"));

var _constants = require("./constants");

var _styledComponents = require("./styled-components");

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
