// baseui/esm/divider/styled-components.js
/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import { styled } from '../styles';
import { SIZE } from './constants';
export var StyledDivider = styled('hr', function (props) {
  var $theme = props.$theme,
      _props$$size = props.$size,
      $size = _props$$size === void 0 ? SIZE.section : _props$$size;

  if ($size === SIZE.cell) {
    return {
      borderTopWidth: '1px',
      borderTopColor: $theme.colors.borderOpaque,
      borderTopStyle: 'solid',
      // base styles
      borderRightWidth: 0,
      borderBottomWidth: 0,
      borderLeftWidth: 0,
      borderRightStyle: 'none',
      borderBottomStyle: 'none',
      borderLeftStyle: 'none'
    };
  }

  if ($size === SIZE.module) {
    return {
      borderTopWidth: '8px',
      borderTopColor: $theme.colors.backgroundTertiary,
      borderTopStyle: 'solid',
      // base styles
      borderRightWidth: 0,
      borderBottomWidth: 0,
      borderLeftWidth: 0,
      borderRightStyle: 'none',
      borderBottomStyle: 'none',
      borderLeftStyle: 'none'
    };
  }

  return {
    borderTopWidth: '2px',
    borderTopColor: $theme.colors.borderOpaque,
    borderTopStyle: 'solid',
    // base styles
    borderRightWidth: 0,
    borderBottomWidth: 0,
    borderLeftWidth: 0,
    borderRightStyle: 'none',
    borderBottomStyle: 'none',
    borderLeftStyle: 'none'
  };
});
StyledDivider.displayName = "StyledDivider";
StyledDivider.displayName = 'StyledDivider';
