// baseui/layout-grid/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.STYLE_VALUES = exports.STYLE = exports.BEHAVIOR = exports.ALIGNMENT = void 0;

var _Object$freeze;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
var BEHAVIOR = Object.freeze({
  fluid: 'fluid',
  fixed: 'fixed'
});
exports.BEHAVIOR = BEHAVIOR;
var ALIGNMENT = Object.freeze({
  start: 'flex-start',
  center: 'center',
  end: 'flex-end'
});
exports.ALIGNMENT = ALIGNMENT;
var STYLE = Object.freeze({
  default: 'default',
  compact: 'compact'
});
exports.STYLE = STYLE;
var STYLE_VALUES = Object.freeze((_Object$freeze = {}, _defineProperty(_Object$freeze, STYLE.default, null), _defineProperty(_Object$freeze, STYLE.compact, {
  columns: [4, 8, 12],
  gutters: [16, 16, 16],
  margins: [16, 24, 24],
  gaps: 0,
  unit: 'px',
  maxWidth: 1280
}), _Object$freeze));
exports.STYLE_VALUES = STYLE_VALUES;
