// baseui/layout-grid/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Unstable_StyledGrid: true,
  Unstable_StyledCell: true,
  Unstable_Grid: true,
  Unstable_Cell: true,
  StyledGrid: true,
  StyledCell: true,
  Grid: true,
  Cell: true
};
Object.defineProperty(exports, "Cell", {
  enumerable: true,
  get: function get() {
    return _cell.default;
  }
});
Object.defineProperty(exports, "Grid", {
  enumerable: true,
  get: function get() {
    return _grid.default;
  }
});
Object.defineProperty(exports, "StyledCell", {
  enumerable: true,
  get: function get() {
    return _styledComponents.StyledCell;
  }
});
Object.defineProperty(exports, "StyledGrid", {
  enumerable: true,
  get: function get() {
    return _styledComponents.StyledGrid;
  }
});
exports.Unstable_StyledGrid = exports.Unstable_StyledCell = exports.Unstable_Grid = exports.Unstable_Cell = void 0;

var _styledComponents = require("./styled-components");

var _grid = _interopRequireDefault(require("./grid"));

var _cell = _interopRequireDefault(require("./cell"));

var _deprecatedComponent = _interopRequireDefault(require("../utils/deprecated-component"));

var _constants = require("./constants");

Object.keys(_constants).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _constants[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _constants[key];
    }
  });
});

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
var componentName = 'Layout Grid (baseui/layout-grid)';
var Unstable_StyledGrid = (0, _deprecatedComponent.default)(_styledComponents.StyledGrid, componentName);
exports.Unstable_StyledGrid = Unstable_StyledGrid;
var Unstable_StyledCell = (0, _deprecatedComponent.default)(_styledComponents.StyledCell, componentName);
exports.Unstable_StyledCell = Unstable_StyledCell;
var Unstable_Grid = (0, _deprecatedComponent.default)(_grid.default, componentName);
exports.Unstable_Grid = Unstable_Grid;
var Unstable_Cell = (0, _deprecatedComponent.default)(_cell.default, componentName);
/** @deprecated use CSSLengthUnit instead. To be removed in future versions.*/

exports.Unstable_Cell = Unstable_Cell;
