// baseui/tag/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Tag: true,
  StyledRoot: true,
  StyledAction: true,
  StyledText: true,
  KIND: true,
  VARIANT: true,
  SIZE: true
};
Object.defineProperty(exports, "KIND", {
  enumerable: true,
  get: function get() {
    return _constants.KIND;
  }
});
Object.defineProperty(exports, "SIZE", {
  enumerable: true,
  get: function get() {
    return _constants.SIZE;
  }
});
Object.defineProperty(exports, "StyledAction", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Action;
  }
});
Object.defineProperty(exports, "StyledRoot", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Root;
  }
});
Object.defineProperty(exports, "StyledText", {
  enumerable: true,
  get: function get() {
    return _styledComponents.Text;
  }
});
Object.defineProperty(exports, "Tag", {
  enumerable: true,
  get: function get() {
    return _tag.default;
  }
});
Object.defineProperty(exports, "VARIANT", {
  enumerable: true,
  get: function get() {
    return _constants.VARIANT;
  }
});

var _tag = _interopRequireDefault(require("./tag"));

var _styledComponents = require("./styled-components");

var _constants = require("./constants");

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
