// baseui/badge/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Badge: true,
  NotificationCircle: true,
  HintDot: true,
  HIERARCHY: true,
  SHAPE: true,
  COLOR: true,
  PLACEMENT: true
};
Object.defineProperty(exports, "Badge", {
  enumerable: true,
  get: function get() {
    return _badge.default;
  }
});
Object.defineProperty(exports, "COLOR", {
  enumerable: true,
  get: function get() {
    return _constants.COLOR;
  }
});
Object.defineProperty(exports, "HIERARCHY", {
  enumerable: true,
  get: function get() {
    return _constants.HIERARCHY;
  }
});
Object.defineProperty(exports, "HintDot", {
  enumerable: true,
  get: function get() {
    return _hintDot.default;
  }
});
Object.defineProperty(exports, "NotificationCircle", {
  enumerable: true,
  get: function get() {
    return _notificationCircle.default;
  }
});
Object.defineProperty(exports, "PLACEMENT", {
  enumerable: true,
  get: function get() {
    return _constants.PLACEMENT;
  }
});
Object.defineProperty(exports, "SHAPE", {
  enumerable: true,
  get: function get() {
    return _constants.SHAPE;
  }
});

var _badge = _interopRequireDefault(require("./badge"));

var _notificationCircle = _interopRequireDefault(require("./notification-circle"));

var _hintDot = _interopRequireDefault(require("./hint-dot"));

var _constants = require("./constants");

var _styledComponents = require("./styled-components");

Object.keys(_styledComponents).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _styledComponents[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _styledComponents[key];
    }
  });
});

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
