// baseui/es/notification/notification.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import * as React from 'react';
import { Toast, TYPE } from '../toast';
import { mergeOverrides } from '../helpers/overrides';
export default class Notification extends React.Component {
  render() {
    const overrides = mergeOverrides({
      Body: {
        style: ({
          $theme
        }) => ({
          marginTop: $theme.sizing.scale600,
          marginRight: $theme.sizing.scale600,
          marginBottom: $theme.sizing.scale600,
          marginLeft: $theme.sizing.scale600
        })
      }
    }, this.props.overrides);
    return /*#__PURE__*/React.createElement(Toast, _extends({
      "data-baseweb": "notification"
    }, this.props, {
      overrides: overrides
    }));
  }

}

_defineProperty(Notification, "defaultProps", {
  notificationType: TYPE.inline,
  closeable: false
});
