// baseui/es/progress-steps/progress-steps.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import * as React from 'react';
import { getOverrides } from '../helpers/overrides';
import { StyledProgressSteps } from './styled-components';

function ProgressSteps({
  overrides = {},
  current,
  children
}) {
  const [Root, rootProps] = getOverrides(overrides.Root, StyledProgressSteps);
  const numChildren = React.Children.count(children); // eslint-disable-next-line @typescript-eslint/no-explicit-any

  const modifiedChildren = React.Children.map(children, (child, index) => {
    if (!child) return;
    const childOverrides = child.props.overrides || {};
    const isActive = child.props.isActive !== undefined ? child.props.isActive : index === current;
    return /*#__PURE__*/React.cloneElement(child, {
      isLast: index === numChildren - 1,
      isCompleted: index < current,
      isActive,
      step: index + 1,
      overrides: { ...overrides,
        Root: overrides.StepRoot,
        ...childOverrides
      }
    });
  });
  return /*#__PURE__*/React.createElement(Root, _extends({
    "data-baseweb": "progress-steps"
  }, rootProps), modifiedChildren);
}

ProgressSteps.defaultProps = {
  current: 0
};
export default ProgressSteps;
