// baseui/map-marker/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.xxSmallPinheadDimension = exports.xSmallPinheadDimension = exports.dragShadowWidth = exports.dragShadowMarginTop = exports.dragShadowHeight = exports.anchorSize = exports.PINHEAD_TYPES = exports.PINHEAD_SIZES_SHAPES = exports.PINHEAD_DIMENSIONS = exports.NEEDLE_SIZES = exports.NEEDLE_HEIGHTS = exports.LOCATION_PUCK_TYPES = exports.LOCATION_PUCK_SIZES = exports.LABEL_SIZES = exports.LABEL_ENHANCER_POSITIONS = exports.KIND = exports.FLOATING_MARKER_SIZES = exports.FLOATING_MARKER_ANCHOR_TYPES = exports.FLOATING_MARKER_ANCHOR_POSITIONS = exports.EARNER_LOCATION_PUCK_CORE_SCALES = exports.BADGE_ENHANCER_STYLES = exports.BADGE_ENHANCER_SIZES = exports.BADGE_ENHANCER_POSITIONS = exports.BADGE_ENHANCER_CONTENT_SIZE = void 0;

var _Object$freeze, _Object$freeze2, _LABEL_SIZES, _xSmallPinheadBadge, _PINHEAD_SIZES_SHAPES, _PINHEAD_SIZES_SHAPES2, _PINHEAD_SIZES_SHAPES3, _BADGE_ENHANCER_POSIT, _BADGE_ENHANCER_STYLE, _BADGE_ENHANCER_CONTE, _Object$freeze3;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
var NEEDLE_SIZES = Object.freeze({
  none: 'none',
  short: 'short',
  medium: 'medium',
  tall: 'tall'
});
exports.NEEDLE_SIZES = NEEDLE_SIZES;
var NEEDLE_HEIGHTS = Object.freeze((_Object$freeze = {}, _defineProperty(_Object$freeze, NEEDLE_SIZES.none, 0), _defineProperty(_Object$freeze, NEEDLE_SIZES.short, 4), _defineProperty(_Object$freeze, NEEDLE_SIZES.medium, 12), _defineProperty(_Object$freeze, NEEDLE_SIZES.tall, 20), _Object$freeze));
exports.NEEDLE_HEIGHTS = NEEDLE_HEIGHTS;
var PINHEAD_SIZES_SHAPES = Object.freeze({
  xxSmallCircle: 'xx-small-circle',
  xxSmallSquare: 'xx-small-square',
  xSmallCircle: 'x-small-circle',
  xSmallSquare: 'x-small-square',
  small: 'small',
  medium: 'medium',
  large: 'large'
});
exports.PINHEAD_SIZES_SHAPES = PINHEAD_SIZES_SHAPES;
var xSmallPinheadDimension = {
  height: 16,
  icon: 4
};
exports.xSmallPinheadDimension = xSmallPinheadDimension;
var xxSmallPinheadDimension = {
  height: 8,
  icon: 4
};
exports.xxSmallPinheadDimension = xxSmallPinheadDimension;
var PINHEAD_DIMENSIONS = Object.freeze((_Object$freeze2 = {}, _defineProperty(_Object$freeze2, PINHEAD_SIZES_SHAPES.xxSmallSquare, xxSmallPinheadDimension), _defineProperty(_Object$freeze2, PINHEAD_SIZES_SHAPES.xxSmallCircle, xxSmallPinheadDimension), _defineProperty(_Object$freeze2, PINHEAD_SIZES_SHAPES.xSmallSquare, xSmallPinheadDimension), _defineProperty(_Object$freeze2, PINHEAD_SIZES_SHAPES.xSmallCircle, xSmallPinheadDimension), _defineProperty(_Object$freeze2, PINHEAD_SIZES_SHAPES.small, {
  height: 24,
  icon: 16
}), _defineProperty(_Object$freeze2, PINHEAD_SIZES_SHAPES.medium, {
  height: 36,
  icon: 16
}), _defineProperty(_Object$freeze2, PINHEAD_SIZES_SHAPES.large, {
  height: 48,
  icon: 24
}), _Object$freeze2));
exports.PINHEAD_DIMENSIONS = PINHEAD_DIMENSIONS;
var PINHEAD_TYPES = Object.freeze({
  floating: 'floating',
  fixed: 'fixed'
});
exports.PINHEAD_TYPES = PINHEAD_TYPES;
var FLOATING_MARKER_SIZES = Object.freeze({
  small: 'small',
  medium: 'medium',
  large: 'large'
});
exports.FLOATING_MARKER_SIZES = FLOATING_MARKER_SIZES;
var FLOATING_MARKER_ANCHOR_POSITIONS = Object.freeze({
  none: 'none',
  topLeft: 'top-left',
  topRight: 'top-right',
  bottomRight: 'bottom-right',
  bottomLeft: 'bottom-left'
});
exports.FLOATING_MARKER_ANCHOR_POSITIONS = FLOATING_MARKER_ANCHOR_POSITIONS;
var FLOATING_MARKER_ANCHOR_TYPES = Object.freeze({
  circle: 'circle',
  square: 'square',
  xxSmallCircle: 'xx-small-circle',
  xxSmallSquare: 'xx-small-square'
});
exports.FLOATING_MARKER_ANCHOR_TYPES = FLOATING_MARKER_ANCHOR_TYPES;
var dragShadowHeight = 4;
exports.dragShadowHeight = dragShadowHeight;
var dragShadowMarginTop = 6;
exports.dragShadowMarginTop = dragShadowMarginTop;
var dragShadowWidth = 6;
exports.dragShadowWidth = dragShadowWidth;
var anchorSize = 16;
exports.anchorSize = anchorSize;
var labelSmall = 'LabelSmall';
var LABEL_SIZES = (_LABEL_SIZES = {}, _defineProperty(_LABEL_SIZES, PINHEAD_SIZES_SHAPES.xxSmallCircle, labelSmall), _defineProperty(_LABEL_SIZES, PINHEAD_SIZES_SHAPES.xxSmallSquare, labelSmall), _defineProperty(_LABEL_SIZES, PINHEAD_SIZES_SHAPES.xSmallCircle, labelSmall), _defineProperty(_LABEL_SIZES, PINHEAD_SIZES_SHAPES.xSmallSquare, labelSmall), _defineProperty(_LABEL_SIZES, PINHEAD_SIZES_SHAPES.small, labelSmall), _defineProperty(_LABEL_SIZES, PINHEAD_SIZES_SHAPES.medium, 'LabelMedium'), _defineProperty(_LABEL_SIZES, PINHEAD_SIZES_SHAPES.large, 'LabelLarge'), _LABEL_SIZES);
exports.LABEL_SIZES = LABEL_SIZES;
var LABEL_ENHANCER_POSITIONS = Object.freeze({
  none: 'none',
  top: 'top',
  left: 'left',
  right: 'right',
  bottom: 'bottom'
});
exports.LABEL_ENHANCER_POSITIONS = LABEL_ENHANCER_POSITIONS;
var BADGE_ENHANCER_SIZES = Object.freeze({
  none: 'none',
  xSmall: 'x-small',
  small: 'small',
  mediumText: 'medium-text',
  mediumIcon: 'medium-icon'
});
exports.BADGE_ENHANCER_SIZES = BADGE_ENHANCER_SIZES;
var xSmallPinheadBadge = (_xSmallPinheadBadge = {}, _defineProperty(_xSmallPinheadBadge, BADGE_ENHANCER_SIZES.none, null), _defineProperty(_xSmallPinheadBadge, BADGE_ENHANCER_SIZES.xSmall, {
  x: -5,
  y: -3
}), _defineProperty(_xSmallPinheadBadge, BADGE_ENHANCER_SIZES.small, null), _defineProperty(_xSmallPinheadBadge, BADGE_ENHANCER_SIZES.mediumText, null), _defineProperty(_xSmallPinheadBadge, BADGE_ENHANCER_SIZES.mediumIcon, null), _xSmallPinheadBadge);
var BADGE_ENHANCER_POSITIONS = (_BADGE_ENHANCER_POSIT = {}, _defineProperty(_BADGE_ENHANCER_POSIT, PINHEAD_SIZES_SHAPES.xxSmallSquare, null), _defineProperty(_BADGE_ENHANCER_POSIT, PINHEAD_SIZES_SHAPES.xxSmallCircle, null), _defineProperty(_BADGE_ENHANCER_POSIT, PINHEAD_SIZES_SHAPES.xSmallSquare, xSmallPinheadBadge), _defineProperty(_BADGE_ENHANCER_POSIT, PINHEAD_SIZES_SHAPES.xSmallCircle, xSmallPinheadBadge), _defineProperty(_BADGE_ENHANCER_POSIT, PINHEAD_SIZES_SHAPES.small, (_PINHEAD_SIZES_SHAPES = {}, _defineProperty(_PINHEAD_SIZES_SHAPES, BADGE_ENHANCER_SIZES.none, null), _defineProperty(_PINHEAD_SIZES_SHAPES, BADGE_ENHANCER_SIZES.xSmall, {
  x: -7,
  y: -1
}), _defineProperty(_PINHEAD_SIZES_SHAPES, BADGE_ENHANCER_SIZES.small, {
  x: -8,
  y: -8
}), _defineProperty(_PINHEAD_SIZES_SHAPES, BADGE_ENHANCER_SIZES.mediumText, null), _defineProperty(_PINHEAD_SIZES_SHAPES, BADGE_ENHANCER_SIZES.mediumIcon, null), _PINHEAD_SIZES_SHAPES)), _defineProperty(_BADGE_ENHANCER_POSIT, PINHEAD_SIZES_SHAPES.medium, (_PINHEAD_SIZES_SHAPES2 = {}, _defineProperty(_PINHEAD_SIZES_SHAPES2, BADGE_ENHANCER_SIZES.none, null), _defineProperty(_PINHEAD_SIZES_SHAPES2, BADGE_ENHANCER_SIZES.xSmall, {
  x: -9,
  y: 1
}), _defineProperty(_PINHEAD_SIZES_SHAPES2, BADGE_ENHANCER_SIZES.small, {
  x: -10,
  y: -4
}), _defineProperty(_PINHEAD_SIZES_SHAPES2, BADGE_ENHANCER_SIZES.mediumText, {
  x: -12,
  y: -8
}), _defineProperty(_PINHEAD_SIZES_SHAPES2, BADGE_ENHANCER_SIZES.mediumIcon, {
  x: -12,
  y: -8
}), _PINHEAD_SIZES_SHAPES2)), _defineProperty(_BADGE_ENHANCER_POSIT, PINHEAD_SIZES_SHAPES.large, (_PINHEAD_SIZES_SHAPES3 = {}, _defineProperty(_PINHEAD_SIZES_SHAPES3, BADGE_ENHANCER_SIZES.none, null), _defineProperty(_PINHEAD_SIZES_SHAPES3, BADGE_ENHANCER_SIZES.xSmall, {
  x: -11,
  y: 3
}), _defineProperty(_PINHEAD_SIZES_SHAPES3, BADGE_ENHANCER_SIZES.small, null), _defineProperty(_PINHEAD_SIZES_SHAPES3, BADGE_ENHANCER_SIZES.mediumText, {
  x: -14,
  y: -6
}), _defineProperty(_PINHEAD_SIZES_SHAPES3, BADGE_ENHANCER_SIZES.mediumIcon, {
  x: -14,
  y: -6
}), _PINHEAD_SIZES_SHAPES3)), _BADGE_ENHANCER_POSIT);
exports.BADGE_ENHANCER_POSITIONS = BADGE_ENHANCER_POSITIONS;
var BADGE_ENHANCER_STYLES = (_BADGE_ENHANCER_STYLE = {}, _defineProperty(_BADGE_ENHANCER_STYLE, BADGE_ENHANCER_SIZES.none, {}), _defineProperty(_BADGE_ENHANCER_STYLE, BADGE_ENHANCER_SIZES.xSmall, {
  height: '8px',
  width: '8px',
  borderRadius: '50%'
}), _defineProperty(_BADGE_ENHANCER_STYLE, BADGE_ENHANCER_SIZES.small, {
  height: '16px',
  width: '16px',
  borderRadius: '50%',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  fontSize: '10px'
}), _defineProperty(_BADGE_ENHANCER_STYLE, BADGE_ENHANCER_SIZES.mediumText, {
  borderRadius: '10px',
  height: '20px',
  padding: '0px 8px',
  display: 'flex',
  alignItems: 'center',
  boxSizing: 'border-box'
}), _defineProperty(_BADGE_ENHANCER_STYLE, BADGE_ENHANCER_SIZES.mediumIcon, {
  height: '20px',
  width: '20px',
  borderRadius: '50%',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  boxSizing: 'border-box'
}), _BADGE_ENHANCER_STYLE);
exports.BADGE_ENHANCER_STYLES = BADGE_ENHANCER_STYLES;
var BADGE_ENHANCER_CONTENT_SIZE = (_BADGE_ENHANCER_CONTE = {}, _defineProperty(_BADGE_ENHANCER_CONTE, BADGE_ENHANCER_SIZES.none, 0), _defineProperty(_BADGE_ENHANCER_CONTE, BADGE_ENHANCER_SIZES.xSmall, 0), _defineProperty(_BADGE_ENHANCER_CONTE, BADGE_ENHANCER_SIZES.small, 10), _defineProperty(_BADGE_ENHANCER_CONTE, BADGE_ENHANCER_SIZES.mediumText, 12), _defineProperty(_BADGE_ENHANCER_CONTE, BADGE_ENHANCER_SIZES.mediumIcon, 12), _BADGE_ENHANCER_CONTE);
exports.BADGE_ENHANCER_CONTENT_SIZE = BADGE_ENHANCER_CONTENT_SIZE;
var KIND = Object.freeze({
  default: 'default',
  accent: 'accent',
  negative: 'negative'
});
exports.KIND = KIND;
var LOCATION_PUCK_SIZES = Object.freeze({
  small: 'small',
  medium: 'medium',
  large: 'large'
});
exports.LOCATION_PUCK_SIZES = LOCATION_PUCK_SIZES;
var LOCATION_PUCK_TYPES = Object.freeze({
  consumer: 'consumer',
  earner: 'earner'
});
exports.LOCATION_PUCK_TYPES = LOCATION_PUCK_TYPES;
var EARNER_LOCATION_PUCK_CORE_SCALES = Object.freeze((_Object$freeze3 = {}, _defineProperty(_Object$freeze3, LOCATION_PUCK_SIZES.small, '50%'), _defineProperty(_Object$freeze3, LOCATION_PUCK_SIZES.medium, '75%'), _defineProperty(_Object$freeze3, LOCATION_PUCK_SIZES.large, '100%'), _Object$freeze3));
exports.EARNER_LOCATION_PUCK_CORE_SCALES = EARNER_LOCATION_PUCK_CORE_SCALES;
