// baseui/badge/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SHAPE = exports.ROLE = exports.PLACEMENT = exports.HIERARCHY = exports.COLOR = void 0;

/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
var HIERARCHY = Object.freeze({
  primary: 'primary',
  secondary: 'secondary'
});
exports.HIERARCHY = HIERARCHY;
var SHAPE = Object.freeze({
  pill: 'pill',
  rectangle: 'rectangle'
});
exports.SHAPE = SHAPE;
var COLOR = Object.freeze({
  accent: 'accent',
  primary: 'primary',
  positive: 'positive',
  negative: 'negative',
  warning: 'warning'
});
exports.COLOR = COLOR;
var PLACEMENT = Object.freeze({
  topLeft: 'topLeft',
  topRight: 'topRight',
  bottomRight: 'bottomRight',
  bottomLeft: 'bottomLeft',
  topLeftEdge: 'topLeftEdge',
  topEdge: 'topEdge',
  topRightEdge: 'topRightEdge',
  bottomRightEdge: 'bottomRightEdge',
  bottomEdge: 'bottomEdge',
  bottomLeftEdge: 'bottomLeftEdge',
  leftTopEdge: 'leftTopEdge',
  rightTopEdge: 'rightTopEdge',
  rightBottomEdge: 'rightBottomEdge',
  leftBottomEdge: 'leftBottomEdge'
});
exports.PLACEMENT = PLACEMENT;
var ROLE = Object.freeze({
  badge: 'badge',
  notificationCircle: 'notificationCircle',
  hintDot: 'hintDot'
});
exports.ROLE = ROLE;
